README.md
# flymake-clippy

[![MELPA](https://melpa.org/packages/flymake-clippy-badge.svg)](https://melpa.org/#/flymake-clippy)

A Flymake backend for [Clippy](https://doc.rust-lang.org/stable/clippy/index.html), the Rust linter.

## Instructions

You probably want to install [rust-mode](https://github.com/rust-lang/rust-mode) first.

Install from [MELPA](https://melpa.org/#/getting-started):

``` elisp
(use-package flymake-clippy
  :hook (rust-mode . flymake-clippy-setup-backend))
```

Alternatively, clone the repo and update your load path:

```
git clone https://git.sr.ht/~mgmarlow/flymake-clippy /path/to/flymake-clippy
```

```elisp
(add-to-list 'load-path "/path/to/flymake-clippy")
(require 'flymake-clippy)
```

### Eglot users

Eglot users require [a little extra setup](https://github.com/joaotavora/eglot/issues/268) to enable running multiple Flymake backends simultaneously. Add the following to your Emacs config:

```elisp
;; Instruct Eglot to stop managing Flymake
(add-to-list 'eglot-stay-out-of 'flymake)

;; Manually re-enable Eglot's Flymake backend
(defun manually-activate-flymake ()
  (add-hook 'flymake-diagnostic-functions #'eglot-flymake-backend nil t)
  (flymake-mode 1))

(add-hook 'eglot-managed-mode-hook #'manually-activate-flymake nil t)
```

(Nb. prior to eglot 1.6, this hook was called `eglot--managed-mode-hook)

You can confirm that Flymake is running correctly by opening up a Rust buffer and examining `flymake-running-backends':

```
M-x flymake-running-backends

Running backends: flymake-clippy-backend, eglot-flymake-backend
```

### Complete eglot + rust-mode + use-package example

With Eglot 1.6+:

```elisp
(use-package rust-mode
  :ensure t)

(use-package flymake-clippy
  :hook (rust-mode . flymake-clippy-setup-backend))

(defun manually-activate-flymake ()
  (add-hook 'flymake-diagnostic-functions #'eglot-flymake-backend nil t)
  (flymake-mode 1))

(use-package eglot
  :ensure t
  :hook ((rust-mode . eglot-ensure)
         (eglot-managed-mode . manually-activate-flymake))
  :config
  (add-to-list 'eglot-stay-out-of 'flymake))
```

## Contributing

Please direct bug reports or patches to the [the mailing list](https://lists.sr.ht/~mgmarlow/public-inbox).

## License

Licensed under [GPL-3.0](./LICENSE).


CHANGELOG.org
* flymake-clippy changelog

** 1.1.0 (11/02/23)

- Add ~flymake-clippy-bin-args~. You can now pass arguments to ~cargo
  clippy~ via the ~flymake-clippy-bin-args~ custom variable. Kudos to
  Jens Östlund for this feature!


clippy-flymake-test.el
;;; clippy-flymake-test.el   -*- lexical-binding: t; -*-

(require 'clippy-flymake)
(require 'ert)

(defun run-regexp ()
  (set-match-data nil)
  (search-forward-regexp (clippy-flymake--build-regexp) nil t)
  (list (match-string 1)
        (match-string 2)
        (match-string 3)))

(ert-deftest clippy-test-regexp ()
  "Tests regexp matches diagnostic information."
  (should (equal (with-temp-buffer
                   (insert-file-contents "./test/fixture.txt")
                   (run-regexp))
                 '("warning: unused variable: `user`" "src/database/foo.rs" "42")))
  (should (equal (with-temp-buffer
                   (insert-file-contents "./test/fixture.txt")
                   (run-regexp)
                   (run-regexp))
                 '("warning: using `clone` on type `Status` which implements the `Copy` trait" "src/foo.rs" "31")))
  (should (equal (with-temp-buffer
                   (insert-file-contents "./test/fixture.txt")
                   (run-regexp)
                   (run-regexp)
                   (run-regexp))
                 '("warning: unused variable: `user`" "src/foobar/user.rs" "42"))))


flymake-clippy-test.el
;;; flymake-clippy-test.el   -*- lexical-binding: t; -*-

(require 'flymake-clippy)
(require 'ert)

(defmacro with-text (text &rest body)
  `(with-temp-buffer
    (insert ,text)
    (goto-char 0)
    ,@body))

(defun run-regexp ()
  (set-match-data nil)
  (search-forward-regexp (flymake-clippy--build-regexp) nil t)
  (list (match-string 1)
        (match-string 2)
        (match-string 3)))

(defvar warning-text
"warning: unused variable: `user`
  --> src/database/foo.rs:42:9
   |
42 |         user: &User,
   |         ^^^^ help: if this is intentional, prefix it with an underscore: `_user`
   |
   = note: `#[warn(unused_variables)]` on by default")

(ert-deftest clippy-test-warnings ()
  "Tests warning diagnostics."
  (should (equal (with-text warning-text (run-regexp))
                 '("warning: unused variable: `user`" "src/database/foo.rs" "42"))))

(defvar error-text
"error: expected one of `!` or `::`, found foobar
  --> src/main.rs:20:9")

(defvar error-text-with-error-num
"error[E0407]: method `build_string` is not defined
  --> src/features.rs:106:5")

(ert-deftest clippy-test-errors ()
  "Tests error diagnostics."
  (should (equal (with-text error-text (run-regexp))
                 '("error: expected one of `!` or `::`, found foobar" "src/main.rs" "20")))
  (should (equal (with-text error-text-with-error-num (run-regexp))
                 '("error[E0407]: method `build_string` is not defined" "src/features.rs" "106"))))

(defvar multiline-text
"warning: unused variable: `user`
  --> src/database/foo.rs:42:9
   |
42 |         user: &User,
   |         ^^^^ help: if this is intentional, prefix it with an underscore: `_user`
   |
   = note: `#[warn(unused_variables)]` on by default

error: expected one of `!` or `::`, found foobar
  --> src/main.rs:20:9")

(ert-deftest clippy-test-multiline ()
  "Tests consecutive `run-regexp' calls."
  (should (equal (with-text multiline-text (run-regexp))
                 '("warning: unused variable: `user`" "src/database/foo.rs" "42")))
  (should (equal (with-text multiline-text (run-regexp) (run-regexp))
                 '("error: expected one of `!` or `::`, found foobar" "src/main.rs" "20"))))


flymake-clippy.el
;;; flymake-clippy.el --- Flymake backend for Clippy  -*- lexical-binding: t; -*-

;; Copyright (C) 2023  Graham Marlow

;; Author: Graham Marlow <info@mgmarlow.com>
;; Keywords: tools
;; URL: https://sr.ht/~mgmarlow/flymake-clippy/
;; Version: 1.1.0
;; Package-Requires: ((emacs "26.1"))

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <https://www.gnu.org/licenses/>.

;;; Commentary:

;; Flymake backend for Clippy, the Rust linter.
;;
;; (use-package flymake-clippy
;;   :hook (rust-mode . flymake-clippy-setup-backend))

;;; Code:

(require 'cl-lib)

(defgroup flymake-clippy nil
  "Flymake backend for Clippy."
  :group 'programming)

(defcustom flymake-clippy-bin-args nil
  "A list of arguments passed to the `cargo clippy' command.

For example, a value of `(\"--\" \"-W\" \"clippy::pedantic\")'
results in the command `cargo clippy -- -W clippy::pedantic'."
  :type '(repeat string)
  :group 'flymake-clippy)

;; Capture group source example:
;; "warning: ..."
;;    --> src/filename.rs
;; 98 | ...
(defun flymake-clippy--build-regexp ()
  "Regexp for Clippy output."
  (rx (seq line-start
           ;; Message
           (group (or "warning" "error")
                  (zero-or-one (seq "[" (repeat 5 alphanumeric) "]"))
                  ":"
                  (zero-or-more nonl))
           "\n"
           (zero-or-more nonl)
           "--> "
           ;; File
           (group
            (zero-or-more nonl))
           ":"
           ;; Line
           (group
            (one-or-more
             (any "0-9")))
           ":"
           ;; Col
           (group
            (one-or-more
             (any "0-9")))
           line-end)))

(defun flymake-clippy--bin ()
  (append (list "cargo" "clippy") flymake-clippy-bin-args))

(defvar-local flymake-clippy--proc nil
  "Clippy subprocess object, used to ensure obsolete processes aren't reused.")

(defun flymake-clippy-backend (report-fn &rest _args)
  "Flymake backend for Clippy, the Rust linter.

Calls REPORT-FN with a list of Flymake diagnostics for the
current buffer.

Use `flymake-clippy-setup-backend' to register the backend
with the appropriate Flymake hook."
  (unless (executable-find "cargo")
    (error "Cannot find cargo"))

  (let* ((source (current-buffer))
         (filename (buffer-file-name source)))
    (save-restriction
      (widen)
      (setq flymake-clippy--proc
            (make-process
             :name "flymake-clippy" :noquery t :connection-type 'pipe
             :buffer (generate-new-buffer "*flymake-clippy*")
             :command (flymake-clippy--bin)
             :sentinel
             (lambda (proc _event)
               (when (memq (process-status proc) '(exit signal))
                 (unwind-protect
                     (if (with-current-buffer source (eq proc flymake-clippy--proc))
                         (with-current-buffer (process-buffer proc)
                           (goto-char (point-min))
                           ;; Collect output buffer into diagnostic messages/locations,
                           ;; exposing them via `report-fn'.
                           (cl-loop
                            while (search-forward-regexp
                                   (flymake-clippy--build-regexp)
                                   nil t)
                            for msg = (match-string 1)
                            for sourcefile = (match-string 2)
                            for (beg . end) = (flymake-diag-region
                                               source
                                               (string-to-number (match-string 3)))
                            for type = (if (string-match "^warning" msg)
                                           :warning
                                         :error)
                            when (and sourcefile (string-match-p sourcefile filename))
                            collect (flymake-make-diagnostic source beg end type msg)
                            into diags
                            finally (funcall report-fn diags)))
                       (flymake-log :warning "Canceling obsolete check %s" proc))
                   ;; Cleanup temp buffer.
                   (kill-buffer (process-buffer proc)))))))
      (process-send-region flymake-clippy--proc (point-min) (point-max))
      (process-send-eof flymake-clippy--proc))))

(defun flymake-clippy-setup-backend ()
  "Add `flymake-clippy' to `flymake-diagnostic-functions' hook."
  (add-hook 'flymake-diagnostic-functions #'flymake-clippy-backend nil t))

(provide 'flymake-clippy)

;;; flymake-clippy.el ends here


clippy-flymake.el
;;; clippy-flymake.el --- Flymake backend for Clippy  -*- lexical-binding: t; -*-

;; Copyright (C) 2023  Graham Marlow

;; Author: Graham Marlow <info@mgmarlow.com>
;; Keywords: tools
;; URL: https://sr.ht/~mgmarlow/clippy-flymake/
;; Version: 1.0.0
;; Package-Requires: ((emacs "26.1"))

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <https://www.gnu.org/licenses/>.

;;; Commentary:

;; Flymake backend for Clippy, the Rust linter.
;; https://doc.rust-lang.org/stable/clippy/index.html

;;; Code:

(require 'cl-lib)

;; Capture group source example:
;; "warning: ..."
;;    --> src/filename.rs
;; 98 | ...
(defun clippy-flymake--build-regexp ()
  "Create a regular expression to search Clippy warnings for FILENAME."
  (rx (seq line-start
           ;; Message
           (group (or "warning:" "error:")
                  (zero-or-more nonl))
           "\n"
           (zero-or-more nonl)
           "--> "
           ;; File
           (group
            (zero-or-more nonl))
           ":"
           ;; Line
           (group
            (one-or-more
             (any "0-9")))
           ":"
           ;; Col
           (group
            (one-or-more
             (any "0-9")))
           line-end)))

(defvar-local clippy-flymake--proc nil
  "Clippy subprocess object, used to ensure obsolete processes aren't reused.")

(defun clippy-flymake-backend (report-fn &rest _args)
  "Flymake backend for Clippy, the Rust linter.

Calls REPORT-FN with a list of Flymake diagnostics for the
current buffer.

Use `clippy-flymake-setup-backend' to register the backend
with the appropriate Flymake hook."
  (unless (executable-find "cargo")
    (error "Cannot find cargo"))

  (let* ((source (current-buffer))
         (filename (buffer-file-name source)))
    (save-restriction
      (widen)
      (setq clippy-flymake--proc
            (make-process
             :name "clippy-flymake" :noquery t :connection-type 'pipe
             :buffer (generate-new-buffer "*clippy-flymake*")
             :command '("cargo" "clippy")
             :sentinel
             (lambda (proc _event)
               (when (memq (process-status proc) '(exit signal))
                 (unwind-protect
                     (if (with-current-buffer source (eq proc clippy-flymake--proc))
                         (with-current-buffer (process-buffer proc)
                           (goto-char (point-min))
                           ;; Collect output buffer into diagnostic messages/locations,
                           ;; exposing them via `report-fn'.
                           (cl-loop
                            while (search-forward-regexp
                                   (clippy-flymake--build-regexp)
                                   nil t)
                            for msg = (match-string 1)
                            for sourcefile = (match-string 2)
                            for (beg . end) = (flymake-diag-region
                                               source
                                               (string-to-number (match-string 3)))
                            for type = (if (string-match "^warning" msg)
                                           :warning
                                         :error)
                            when (and sourcefile (string-match-p sourcefile filename))
                            collect (flymake-make-diagnostic source beg end type msg)
                            into diags
                            finally (funcall report-fn diags)))
                       (flymake-log :warning "Canceling obsolete check %s" proc))
                   ;; Cleanup temp buffer.
                   (kill-buffer (process-buffer proc)))))))
      (process-send-region clippy-flymake--proc (point-min) (point-max))
      (process-send-eof clippy-flymake--proc))))

(defun clippy-flymake-setup-backend ()
  "Add `clippy-flymake' to `flymake-diagnostic-functions' hook."
  (add-hook 'flymake-diagnostic-functions #'clippy-flymake-backend nil t))

(provide 'clippy-flymake)

;;; clippy-flymake.el ends here
